packages/sdk/src/types/runtime/assets/assets-v2.md

```diff
@@ -0,0 +1,100 @@
+# Asset Unloading
+
+This proposal details the API changes required to allow developers to explicitly request
+that a given asset (material, texture, sound) be unloaded from memory on the clients,
+and unassign all references to them.
+
+Today, assets can be created individually, or created en masse by loading a container
+(glTF) for assets. This proposal does not have separate unload APIs for these two cases.
+Instead, during synchronization, the adapter will examine which assets are loaded at sync
+time, send a container load request if at least one asset from that container is still
+live, then send unloads for all the unused ones. This is processor-inefficient, but
+maintains API focus on the concept of assets.
+
+An alternative to the above would be to modify `AssetManager.loadGltf` to allow the
+developer to specify which assets are loaded. This is technically possible and solves
+the inefficiencies of the above, but requires developer knowledge of glTF internals,
+which may not be feasible. As another alternative, the API could require that
+container-loaded assets be unloaded as a block. This alternative doesn't really add
+any new functionality to the proposal, but reduces multipeer adapter complexity.
+
+## Proposed API Changes
+
+1. `AssetManager.unloadAssets(...assetIds: Asset[]): Promise<void>`
+
+	This method does four things:
+
+	1. It breaks all references to each asset by using the asset's internal reference
+		tracker, which will store references to each actor/asset that uses the asset.
+	2. It waits for the next batch of updates to propagate to the clients, so the assets
+		to be unloaded are not in use during the unload. This wait is done via a new
+		context method that resolves a promise when the next update is done.
+	3. It sends an `unload-assets` message to clients with the asset IDs. This will
+		trigger an `Object.Destroy` in Unity's asset cache, freeing the memory.
+	4. It collects success/failure messages from the client, and propagates any errors
+		from the client unload.
```

Comment:
I think UnloadAssets() is a confusing API, if Asset[] can either be top-level (a preloaded GLTF, or a programmatically created), or owned by a higher level asset (the GLTF owns textures, materials etc). 
I think we should either use unloadGroup() only, or consider having individual assets are all creating a TopLevelAsset or AssetContainer or something like that, and the only thing you can free are asset containers. I think unloadGroup() is simpler

Reply:
I understand and discuss the performance argument, but you think the API design is confusing? From the developer perspective, the only objects that are useful to building an experience are at the asset level. Making them track groups as well seems unnecessarily burdensome to me, given that there's no technical requirement for it.

---
In reply to: [300189061](https://github.com/microsoft/mixed-reality-extension-sdk/pull/369#discussion_r300189061) [](ancestors = 300189061)

Reply:
The design of UnloadAssets goes contrary to the way content managers typically work in game engines. Loads and unloads should be symmetric. The complexity of the API makes it less discoverable, and also a burden on future runtime implementors (not all content loaders will support partial unloads). I'm in favor of the asset container approach, and only supporting unload at the container level.